### PR TITLE
Visitor Video Popup

### PR DESCRIPTION
**Pull request by deniss-mc**
**Jira issue:**
https://glia.atlassian.net/browse/MUIC-648

**Additional info**
There is some sort of visual bug when you hover visitor video over operator video edges that are over operator video are displayed sharp, so edges are now set to sharp to get rid of the issue. (Needs core-SDK team input - or research)

Also sometimes when operator video is created after visitor video (post methods order), visitor video is positioned behind the operator video. Not sure why (might need core-SDK team input on this) - tried to add z-vector and elevation but it still happens. View hierarchy seems to be ignored? (check last image)

**Screenshots:**
Screenshot_1649770369
Screenshot_1649770379
Screenshot_1649770383
Screenshot_1649839170

### CODE REVIEW

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public void setActivity(Activity activity) {
+        this.activity = activity;
+    }
+
+    @Override
+    public void show(VisitorMediaState state) {
+        post(() -> {
+            if (!floatingVisitorVideoContainer.hasVideo()) {
+                floatingVisitorVideoContainer.showVisitorVideo(state.getVideo().createVideoView(activity));
```
Comment (jakob-mc) [inline widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java:55]: pls wrap

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoView.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        pauseVideoStream();
+    }
+
+    public void onDestroy() {
+        releaseVideoStream();
+    }
+
+    public void showVisitorVideo(VideoView newVideoView) {
+        videoView = newVideoView;
+        videoView.setZ(20);
```
Comment (jakob-mc) [inline widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoView.java:50]: dat magic number

**Review by jakob-mc — CHANGES_REQUESTED** [review]

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoView.java (hunk 2 above)
Comment (AndriiHorishniiMOC) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoView.java:50]: Unfortunately, looks like it's the issue in the core SDK.

File: widgetssdk/src/main/java/com/glia/widgets/call/CallView.java [hunk 3]
```diff
@@ -1017,36 +1004,14 @@ private void showOperatorVideo(MediaState operatorMediaState) {
         if (operatorMediaState != null && operatorMediaState.getVideo() != null) {
             Logger.d(TAG, "Starting video operator");
             operatorVideoView = operatorMediaState.getVideo().createVideoView(Utils.getActivity(this.getContext()));
+            operatorVideoView.setZ(10);
```
[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/call/CallView.java:1007]: ZOrderMediaOverlay - elevation order - I will remove everything related to that as it is not working at the moment

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.core.util.Pair;
+
+import com.glia.androidsdk.comms.VisitorMediaState;
+import com.glia.widgets.R;
+import com.glia.widgets.di.Dependencies;
+import com.glia.widgets.view.ViewHelpers;
+
+public class FloatingVisitorVideoCoordinatorLayout extends FrameLayout implements FloatingVisitorVideoContract.View {
```
[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java:19]: It coordinates and holds the floating view, please suggest better naming here :)

File: widgetssdk/src/main/java/com/glia/widgets/call/CallView.java [hunk 5]
```diff
@@ -796,7 +782,8 @@ private void hideUIOnCallEnd() {
         setVisibility(INVISIBLE);
         Activity activity = Utils.getActivity(this.getContext());
         hideOperatorVideo();
-        hideVisitorVideo();
+        floatingVisitorVideoCoordinatorLayout.setVisibility(GONE);
+        floatingVisitorVideoCoordinatorLayout.hide();
```
[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/call/CallView.java:786]: forgot to remove, good catch 👍

File: widgetssdk/src/main/java/com/glia/widgets/call/CallView.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/call/CallView.java]: removed - will add elevation stuff back later if required :)

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoView.java (hunk 2 above)
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoView.java]: removed - will add elevation stuff back later if required :)

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java (hunk 1 above)
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java]: automatic wrapping didn't work here btw. bit weird

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoController.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        addVisitorMediaStateListenerUseCase.execute(this);
+    }
+
+    @Override
+    public void onPause() {
+        removeVisitorMediaStateListenerUseCase.execute(this);
+    }
+
+    @Override
+    public void onDestroy() {
```
[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoController.java:35]: no-op

[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoController.java:35]: part of base controller

[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoController.java:35]: I'll tie it with view tho, hope your error will go away x)

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakob-mc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java:19]: just drop the Coordinator part? It's a nitpick, feel free to ignore.

**Review by jakob-mc — APPROVED** [review]

**Comment by andrews-moc** [comment]
@deniss-mc @jakob-mc Quick update on the overlay issue. We are working on it and will update you with our ideas/solutions today.

File: widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrews-moc) [reply widgetssdk/src/main/java/com/glia/widgets/view/floatingvisitorvideoview/FloatingVisitorVideoCoordinatorLayout.java:19]: Agree, I'd also go with "FloatingVisitorVideo" name.

**Review by dukhovnyi — APPROVED** [review]

**Pull request merged**